Register audio and image models consistently in admin panel

dhruvkb opened the pull request:
## Fixes
<!-- If PR doesn't fully resolve the issue, replace 'Fixes' below with 'Related to'. -->
<!-- If there is no issue being resolved, please consider opening one before creating this pull request. -->
Fixes #919 by @krysal 

## Description
<!-- Concisely describe what the pull request does. -->
<!-- Add screenshots, videos, or other media to show the problem and the solution when appropriate. -->
This PR
- ensures that images and audio have all reporting models registered in the admin panel
- adds checks to ensure non-existent identifiers cannot be reported
- distributes concerns from the central `_Report` models to the `Deleted_` and `Mature_` models
- ensures that the changes to mature status are synced to Elasticsearch
- surface `mature` field in the details endpoint for WordPress/openverse-frontend#1822
- limits admin functionality to encourage correct usage of reporting flows
- adds documentation where it was lacking
- implements general refactors and makes code quality improvements

## Testing Instructions
<!-- Give steps for the reviewer to verify that this PR fixes the problem; or delete this section entirely. -->
0. Check out the PR and run all services.
1. Visit the Django admin panel.
2. See new models present.

## Checklist
<!-- Replace  the [ ] with [x] to check the boxes. -->
- [x] My pull request has a descriptive title (not a vague title like `Update index.md`).
- [x] My pull request targets the *default* branch of the repository (`main`) or a parent feature branch.
- [x] My commit messages follow [best practices][best_practices].
- [x] My code follows the established code style of the repository.
- [x] I added or updated tests for the changes I made (if applicable).
- [x] I added or updated documentation (if applicable).
- [x] I tried running the project locally and verified that there are no visible errors.

[best_practices]:https://git-scm.com/book/en/v2/Distributed-Git-Contributing-to-a-Project#_commit_guidelines

## Developer Certificate of Origin
<!-- You must read and understand the following attestation. -->

<details>
<summary>Developer Certificate of Origin</summary>

```
Developer Certificate of Origin
Version 1.1

Copyright (C) 2004, 2006 The Linux Foundation and its contributors.
1 Letterman Drive
Suite D4700
San Francisco, CA, 94129

Everyone is permitted to copy and distribute verbatim copies of this
license document, but changing it is not allowed.


Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
```

</details>

krysal reviewed (dismissed):
So nice how clean it looks the generalized code ✨

However, I can register an image uuid as a MatureAudio and vice-versa, can we prevent that from happening? Strangely the image/audio still shows as not mature after registration but that can be a bug for another issue.

sarayourfriend commented on api/catalog/api/models/media.py:
```diff
@@ -219,34 +206,95 @@ def save(self, *args, **kwargs):
     class Meta:
         abstract = True
 
+    def _update_es(self, raise_errors: bool) -> models.Model:
```
Nice idea!

sarayourfriend commented on api/catalog/api/models/audio.py:
```diff
@@ -175,6 +176,10 @@ class Audio(AudioFileMixin, AbstractMedia):
         help_text="JSON describing alternative files for this audio.",
     )
 
+    @property
+    def mature(self) -> bool:
+        return MatureAudio.objects.filter(identifier=self.identifier).exists()
+
```
~Also occurred to me, could this be moved into the `AbstractMedia` class, given it'll probably need to be implemented for every media type? A `mature_class` property on the implementing class would make this nicely generic.~

Update: Never mind, I see that this is probably not easily or ergonomically possible now due to `Mature_` itself requiring the media model class defined on it. The references would be circular.

dhruvkb replied:
Based on my understanding, the `@cached_property` decorator wouldn't not do much here, since that it's lifetime is bound to the lifetime of the model instance?

Django has a decorator with the same name, and their [docs](https://docs.djangoproject.com/en/4.1/ref/utils/#django.utils.functional.cached_property) state that the cache only persists as long as the instance is in memory.

sarayourfriend reviewed (commented):
LGTM so far. I'm checking out the code now to run it locally. I just have one suggestion about the use of `@property` and I/O. I deeply appreciate the unit tests, super helpful :pray:

sarayourfriend commented:
Hmmm, I'm not able to get my local API to spin up. It hangs waiting for the image index to be ready. I'll debug this a bit more later today so I can finish this review.

AetherUnbound commented:
> Hmmm, I'm not able to get my local API to spin up. It hangs waiting for the image index to be ready. I'll debug this a bit more later today so I can finish this review.

FWIW a `just recreate` worked for me! This might be an issue with the ingestion server, RE #417

dhruvkb commented on api/catalog/api/models/image.py:
```diff
@@ -48,41 +49,43 @@ class Image(ImageFileMixin, AbstractMedia):
     class Meta(AbstractMedia.Meta):
         db_table = "image"
 
+    @property
+    def mature(self) -> bool:
+        return MatureImage.objects.filter(identifier=self.identifier).exists()
```
Based on my understanding, the `@cached_property` decorator wouldn't not do much here, since that it's lifetime is bound to the lifetime of the model instance?

Django has a decorator with the same name, and their [docs](https://docs.djangoproject.com/en/4.1/ref/utils/#django.utils.functional.cached_property) state that the cache only persists as long as the instance is in memory.

AetherUnbound commented on api/test/unit/models/media_report_test.py:
```diff
@@ -0,0 +1,138 @@
+@pytest.fixture
+def media_obj():
+    def _get_media_obj(type: MediaType):
+        klass = Audio if type == "audio" else Image
```
This could potentially be a `dict::get` call, which might make it easier to add future media types:
```suggestion
        klass = {
            "audio": Audio,
            "image": Image,
        }.get(type)
```

AetherUnbound commented on api/test/unit/models/media_report_test.py:
```diff
@@ -0,0 +1,138 @@
+@pytest.mark.parametrize(
+    "media_type, report_class", [("image", ImageReport), ("audio", AudioReport)]
+)
+@pytest.mark.parametrize("reason", [DMCA, MATURE])
```
Should we add `OTHER` here as a possible reason too?

dhruvkb replied:
Implemented in fafcdf8.

AetherUnbound commented on api/test/unit/models/media_report_test.py:
```diff
@@ -0,0 +1,138 @@
+@pytest.mark.parametrize(
+    "media_type, report_class", [("image", ImageReport), ("audio", AudioReport)]
+)
+@pytest.mark.parametrize("reason", [DMCA, MATURE])
```
This is only used in two tests, but we could potentially set this to a variable for use in multiple tests:

```python
reason_params = pytest.mark.parametrize("reason", [DMCA, MATURE, OTHER])

...

@reason_params
def test_foo()
```

dhruvkb replied:
Implemented in fafcdf8.

AetherUnbound reviewed (approved):
This is fantastic! The tests are excellent 👩🏼‍🍳 💯 I was able to test adding reports in the Django Admin and it all behaved as expected!

sarayourfriend commented:
> FWIW a just recreate worked for me! This might be an issue with the ingestion server, RE https://github.com/WordPress/openverse-api/issues/417

Nah, for me it's because my disk is too full and ES won't let you add or modify indexes if your disk is above 95% full :sweat_smile: Working on moving some things around to fix this issue. I have been successfully ignoring it for a _long_ time.

Should be able to finish reviewing this soon.

dhruvkb commented:
@sarayourfriend if you do `just down -v` that will delete the ES volume and free up space.

sarayourfriend commented:
> @sarayourfriend if you do just down -v that will delete the ES volume and free up space.

Yeah, that didn't work. Based on the logs I was seeing ES was measuring the available disk space on the host device, somehow. My main drive is filled to 97% so I'm moving some things off to external drives now to see if it'll fix the situation.

This is entirely a problem of my own creation and nothing to do with this project specifically. Sorry, just gotta bear with me a moment on this one.

sarayourfriend reviewed (changes requested):
Noticed a couple things. Some are just nice to haves and could be pushed off into separate issues:

1. The plural form of the audio models could be changed to use `audio` instead of `audios`
2. The order of the items in the Django admin could be rearranged to group audio and image related stuff together. Or maybe just moving the reports to the top of the list so that they're easier to find.
3. I am concerned that we actually delete media objects when we deindex them after review. Should we explore "soft deletion" instead? I noticed this because I tried to create a mature audio report on an audio that I had just deleted.
4. Deleted_ and Mature_ admins allow modifying the identifier. I think this should be a read-only field on these models? Changing the identifier on them and saving them has a strange behaviour: it does not modify the existing record but actually creates a new one. We should prevent this entirely by making these models readonly from the admin. (This is why I am requesting changes on this review).

Otherwise this LGTM :grin:

dhruvkb commented:
1. Fixed in 1f5aee8. I had not initially done it because Django requires a migration when you change the plural name so the tradeoff didn't seem worth it.
2. Rearranging is possible but it is significantly complicated and will increase the scope of the PR by a lot. Will create a separate issue and PR to address this.
3. Deleted images are not reindexed because during the data refresh we exclude all identifiers for which we have a `DeletedImage` or `DeletedAudio` model instance.
4. Since the identifier UUID is a primary field changing it from the admin makes Django think you're working with a new instance and doesn't affect the old one. Fixed in 41a5339.

sarayourfriend reviewed (approved):
Wonderful! Great work, Dhruv :rocket:
